ldrshared/clients/messenger/pubsub.py

```diff
@@ -0,0 +1,250 @@
+from concurrent.futures._base import TimeoutError
+import datetime
+import logging
+import json
+import os
+from typing import Union, List
+
+from google.cloud.pubsub_v1 import PublisherClient, SubscriberClient
+from google.cloud.pubsub_v1.subscriber.futures import StreamingPullFuture
+from google.cloud.pubsub_v1.publisher.futures import Future
+from google.pubsub_v1.types import pubsub as pubsub_gapic_types
+from google.cloud.pubsub_v1.subscriber.message import Message
+from tinydb import TinyDB, Query
+
+from .base import BaseMessenger, BaseMessage
+
+STR_TIME = "%b/%d/%Y %H:%M:%S"
+TIME_STAMP_KEY = "_published_time_stamp"
+
+GCP_PROJECT_ID = os.environ.get("GCP_PROJECT_ID", None)
+if not GCP_PROJECT_ID:
+    raise ValueError("You need to set the project id via the env var GCP_PROJECT_ID")
+
+
+# Note: docstrings inherit from base class
+class PubSubMessage(BaseMessage):
+    message: Message
+
+    def get_attribute(self, key: str, default=None):
+        return self.message.attributes.get(key, default)
+
+    def get(self, key=None, default=None) -> Union[dict, str, list]:
+        data: str = self.message.data.decode("utf-8")
+        if key:
+            data_as_dict: dict = json.loads(data)
+            return data_as_dict.get(key, default)
+
+        try:
+            return json.loads(data)
+        except:
+            return data
+
+
+# TODO - will need to be implemented with an actual database
+# before we start subscribing from more than one location
+class Deduplicator:
+    """
+    Uses simple database operations to check if a given message has been
+    received before.
+    """
+
+    def __init__(self, dbname: str, message_retention: dict):
+        self.db = TinyDB(dbname)
+        self.message_retention: datetime.timedelta = datetime.timedelta(
+            days=message_retention["days"] if "days" in message_retention else 0,
```

Comment:
How *pythonic*. I have to say I don't like untyped dictionaries like this, but this is small enough I'm not going to be insistent.

Reply:
It's a little nastier than I usually would but it is temporary "for now" code, I've have added a couple of assertions to lock it down though, easy enough.